### PR TITLE
[BUG] SDK creates a lot of HttpClient causing connection starvation

### PR DESCRIPTION
**Pull request by divyakgupta**
**Describe the bug**
When using the SDK for long-running processes, we've noticed that it's causing Linux workloads to run out of connections (https://github.com/tomkerkhove/promitor/issues/798).

After digging deeper, we thought we were not using the SDK correctly so we've POCed a refactoring where we only authenticate once and re-use it for the rest of the processing (https://github.com/tomkerkhove/promitor/pull/844).

After digging deeper, [@maartenba found that the SDK is creating a lot of HttpClients ](https://github.com/tomkerkhove/promitor/pull/844#issuecomment-577135547) and not re-using them.

Is this on purpose or is this a bug?

***Exception or Stack Trace***
```
System.Net.Http.HttpRequestException: Too many open files in system ---> System.Net.Sockets.SocketException: Too many open files in system
   at System.Net.Sockets.Socket..ctor(AddressFamily addressFamily, SocketType socketType, ProtocolType protocolType)
   at System.Net.Sockets.DualSocketMultipleConnectAsync..ctor(SocketType socketType, ProtocolType protocolType)
   at System.Net.Sockets.Socket.ConnectAsync(SocketType socketType, ProtocolType protocolType, SocketAsyncEventArgs e)
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
```

**To Reproduce**
- Authenticate client
- Have look how many clients are created

**Expected behavior**
Re-use of one HttpClient rather than 70+

**Setup (please complete the following information):**
 - OS: Linux
 - IDE : Visual Studio
 - Version of the Library used: 1.30.0

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X] Bug Description Added
- [X] Repro Steps Added
- [X] Setup information Added

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@divyakgupta__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (divyag). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, or work for Microsoft Open Technologies, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/AesCbcHmacSha2.cs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+                // Compute the tag
+                _tag = new byte[_hmac_key.Length];
+                Array.Copy( _hmac.Hash, _tag, _hmac_key.Length );
+
+                return result;
+            }
+
+            public void Dispose()
+            {
+ 	           Dispose( true );
```
Comment (matt-gibbs) [inline src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/AesCbcHmacSha2.cs]: nit - whitespace.

File: src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/AesKw.cs [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+    {
+        const int BlockSizeInBits  = 64;
+        const int BlockSizeInBytes = BlockSizeInBits >> 3;
+
+        protected static readonly byte[]                   DefaultIv = new byte[] { 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6 };
+        protected static readonly RNGCryptoServiceProvider Rng       = new RNGCryptoServiceProvider();
+
+        protected AesKw( string name ) : base( name )
+        {
+
```
Comment (matt-gibbs) [inline src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/AesKw.cs]: nit - extra blank line

File: src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/NativeMethods.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                                                               [In] ref NCRYPT_PKCS1_PADDING_INFO pPaddingInfo,
+                                                               [In, MarshalAs( UnmanagedType.LPArray )] byte[] pbHashValue,
+                                                               int cbHashValue,
+                                                               [In, MarshalAs( UnmanagedType.LPArray )] byte[] pbSignature,
+                                                               int cbSignature,
+                                                               AsymmetricPaddingMode dwFlags );
+
+        internal static byte[] NewNCryptPublicBlob( RSAParameters rsaParams )
+        {
+            // Builds a BCRYPT_RSAKEY_BLOB strucutre ( http://msdn.microsoft.com/en-us/library/windows/desktop/aa375531(v=vs.85).aspx ).
```
Comment (matt-gibbs) [inline src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/NativeMethods.cs]: really long comments - ideally split into two lines

File: src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/NativeMethods.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            writer.Write( (int)0 );
+            writer.Write( (int)0 );
+            writer.Write( rsaParams.Exponent );
+            writer.Write( rsaParams.Modulus );
+            return data;
+        }
+
+        internal static byte[] NewNCryptPrivateBlob( RSAParameters rsaParams )
+        {
+            // Builds a BCRYPT_RSAKEY_BLOB strucutre ( http://msdn.microsoft.com/en-us/library/windows/desktop/aa375531(v=vs.85).aspx ).
```
Comment (matt-gibbs) [inline src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/NativeMethods.cs]: really long comments - ideally split into two lines

File: src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/Rsa15.cs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                throw new ArgumentException( "key must be an instance of RSACryptoServiceProvider", "key" );
+
+            return new Rsa15Decryptor( csp );
+        }
+
+        /// <summary>
+        /// RSA 15 Decryptor
+        /// </summary>
+        /// <remarks>
+        /// While this class has a reference to an IDisposable object,
```
Comment (matt-gibbs) [inline src/KeyVault/Microsoft.Azure.KeyVault.Extensions/Cryptography/Algorithms/Rsa15.cs]: Does this cause an fxcop warning? Suppression needed?

**Comment by ogail** [comment]
@divyakgupta please address @matt-gibbs comments

**Comment by markcowl** [comment]
@divyakgupta This pull request has had merge conflicts for 5 days.  Closing - please reopen once you have fixed the merge conflicts and addressed the feedback.

**Comment by jsquire** [comment]
Thank you for your contribution.  Closing this out, as there hasn't been recent activity.  Please feel free to reopen if you'd like to continue working on these changes.

**Pull request closed without merging**